### DISCUSSION TITLE
Close : SQL query with SUBSTR and accented letter.

### DISCUSSION THREAD

**Discussion by meteos77** [textbrowser/biblioteq#257] [Q&A] [answered]
SQLite database with authors whose names have accented characters (É)
authors : SEMPÉ ; SÉMONIN

I am looking for the first 3 letters of the author's name
SELECT location AS Type, SUBSTR(author,1,**3**) AS Auteur,accession_number AS Numéro FROM book
Result:
SEM -> 3 characters
SÉ -> 2 characters

SELECT location AS Type, SUBSTR(author,1,**4**) AS Auteur,accession_number AS Numéro FROM book
Result:
SEMP  -> 4 characters
SÉM -> 3 characters

The SQL query seems to have 2 characters for an accented letter.
How to get around the difficulty ?

In advance thank you for the help

**Comment by textbrowser** [comment]
sqlite> select substr(author,1,4) from book;
SÉMO
Mene
Katz
Paar
Auth
Auth
Anil
Roal
Jorg
SEMP
Barn
Oska

**Comment by textbrowser** [comment]
select substr(author,1,3) from book;
SÉM
Men
Kat
Paa
Aut
Aut
Ani
Roa
Jor
SEM
Bar
Osk

**Comment by textbrowser** [comment]
Identical results in BQ.

**Comment by meteos77** [comment]
![PB-SUBSTR](https://user-images.githubusercontent.com/56431039/205448763-fa1c682c-108b-4dd6-9527-ffec2b35f836.png)

**Reply by meteos77** [reply]
Is this coming from my base?
an extract
[Base-TEST-SUBSTR.sqlite.tar.gz](https://github.com/textbrowser/biblioteq/files/10146822/Base-TEST-SUBSTR.sqlite.tar.gz)

**Reply by textbrowser** [reply]
No. I simply modify the authors according to your complaints and execute the statements.

**Reply by textbrowser** [reply]
The results are correct. Perhaps it's your SQLite engine in a French environment mangling the data. I do not have a French environment so I cannot reproduce it. Maybe you need to set the environment in SQLite. UTF-8 for example.

**Reply by textbrowser** [reply]
Good luck. It's Saturday.

**Reply by meteos77** [reply]
thank you for the track
I'll look for it and let you know

**Reply by meteos77** [reply]
bad result under 
LinuxMint 21 (personnal compilation)
Windows10 (version 2022-11-25 official)

**Reply by textbrowser** [reply]
PQ has the following feature.

https://www.postgresql.org/docs/current/unaccent.html

**Reply by meteos77** [reply]
report:
- when creating via the BiblioteQ interface, it works well
- in creation via csv import, I have the problem
![SUBSTR-Etape2](https://user-images.githubusercontent.com/56431039/205454397-d83b57de-b6dd-42e1-bce7-4ca7d9233951.png)

the import file is in utf8
[import-51-documents.csv.tar.gz](https://github.com/textbrowser/biblioteq/files/10147026/import-51-documents.csv.tar.gz)

file import-51-documents.csv  -> import-51-documents.csv: Unicode text, UTF-8 text, with very long lines (1625)

**Reply by textbrowser** [reply]
I'm going to remove the import tool.

**Reply by textbrowser** [reply]
Too many expectations for it. And I'm tired of parsing text.

**Reply by meteos77** [reply]
forget it, I won't put accents on the authors' names

**Reply by textbrowser** [reply]
The preview is identical to the insert. What you see is what you get.

**Reply by textbrowser** [reply]
The insert in the import is parameterized just like the insert into a normal book.

**Reply by textbrowser** [reply]
But something's different.

**Reply by textbrowser** [reply]
Even after forcing UTF-8, the import fails to produce correct entries.

**Reply by textbrowser** [reply]
Alive?

**Reply by textbrowser** [reply]
![Screenshot_20221203_145044](https://user-images.githubusercontent.com/10701156/205459270-b6371987-cca8-432e-9f08-b9078f00a816.png)

**Reply by textbrowser** [reply]
After setting the codec of the input stream to ISO 8859-1. A new widget is necessary.

**Reply by textbrowser** [reply]
Nope. Some other fields are mangled.

**Reply by textbrowser** [reply]
Binding types are fine in the sub-string select. Perhaps your data is faulty. Anyway, I spent 6 hours of my Saturday on this. Done.

**Reply by meteos77** [reply]
thank you for looking into my problem

if it's my data, I'll look precisely from the beginning of the process because I get the records from a french website with an isbn13 list to a unimarc format to convert them to BQ compatible csv for import by a personal python script.

**Reply by textbrowser** [reply]
If one data field is fine and another is not and both fields with respect to BQ are identical, that suggests that it's not BQ.

**Reply by meteos77** [reply]
I have made progress in understanding
![hexa-des-2-caractères](https://user-images.githubusercontent.com/56431039/205507265-f6371374-20ec-4dfe-936c-a29b7b123133.png)


the 2 E's are coded differently
the one that works
![É_correcte](https://user-images.githubusercontent.com/56431039/205507217-10c70f21-1871-4926-8503-f2545a3427ec.png)


the one that doesn't work and the association of 2
![E_en_format_UTF8](https://user-images.githubusercontent.com/56431039/205507235-176d462e-1b84-4705-b0c3-9bf31a93ca75.png)
![CC81-accent](https://user-images.githubusercontent.com/56431039/205507248-419643b0-397b-48b9-b8cd-1093ffcdc616.png)

so on the display it is correct but in the processing it causes problems.

**Reply by textbrowser** [reply]
I can set the codec on the input stream. It corrects author but breaks binding_type. This suggests that the author is incorrect. Perhaps an accented capital E or however you created the CSV file.

**Reply by textbrowser** [reply]
Anyway, this is your responsibility. :P

**Reply by meteos77** [reply]
here I agree with you :-)
it's a detail that at worst will be handled by a bash or python script and not by SQL
but I will look at my data source because it propagates several unimarc exports

**Reply by meteos77** [reply]
I thought that utf8 simplified things

**Comment by textbrowser** [comment]
![Screenshot_20221203_103750](https://user-images.githubusercontent.com/10701156/205449118-fa5c2539-506e-4d22-988e-d105d83cf9e4.png)

**Comment by textbrowser** [comment]
I cannot reproduce this so I cannot correct it.

**Comment by textbrowser** [comment]
Did you perform a SUBSTR function on other accented data? Notice that the results are expected?

**Reply by meteos77** [reply]
in fact it was the first time I used this function

test in my database transferred from the old software, **it's OK**

there it was with the aim of recording the novelties and to make mailing with [glabels](https://github.com/jimevins/glabels-qt) to make the labels of the edge of the books with like base of label:

type of document
3 letter of the author
access number

**Reply by textbrowser** [reply]
So did you test the SUBSTR() function on binding_type for example? I tested it and the results were normal.

I'm curious of this.

SELECT LENGTH(author), author FROM book

**Reply by meteos77** [reply]
![111](https://user-images.githubusercontent.com/56431039/205507851-901a6771-1449-44ac-8cb7-29fb0e573b85.png)

**Reply by textbrowser** [reply]
See? Hidden character.

**Reply by meteos77** [reply]
while searching, I found this site in French it talks about an extension to sqlite : 
https://blog.slucas.fr/blog/sqlite-icu-extension-1/

**Reply by textbrowser** [reply]
You need to start thinking in SQL and tutor your comrades at the library.

**Reply by textbrowser** [reply]
Because I'm going to start covering my ears and chant to myself whenever I receive a blue notification for a "problem."

**Reply by meteos77** [reply]
vous avez de l'humour :-)
you have a sense of humor :-)

**Reply by meteos77** [reply]
It's not a QB problem, it's you who wrote it. so this is a discussion

**Reply by textbrowser** [reply]
I don't have a sense of humor that I'm aware of.

**Reply by meteos77** [reply]
tutor your comrades at the library.

**Comment by textbrowser** [comment]
https://docs.freebsd.org/en/articles/problem-reports/

**Reply by meteos77** [reply]
I have read through the docs, I am not ready to report to BSD

**Reply by textbrowser** [reply]
Not to BSD. :P Problem reporting requires thinking.

**Reply by meteos77** [reply]
I try to find a title that sums up well
I explain my problem.

Maybe you think it's not explanatory enough but at first I thought my SQL query was missing somewhere.

Of course afterwards i wouldn't even have made the request for help.
but I wasn't even aware that the character could be split into 2 parts and still be displayed in the interface.

**Reply by textbrowser** [reply]
Discussions bring people together. Thoughts arise for revisions, developing new structures and paradigms. I sometimes prefer silence.

**Reply by textbrowser** [reply]
Thinking in SQL is like having a vacation without leaving your habitat. Seeing the world with fresh eyes.

**Reply by meteos77** [reply]
I don't know if the expression suits you: silence is golden. Got it -
I'll go on vacation to SQL country then.


Have a nice Sunday

**Reply by textbrowser** [reply]
Silence is golden sometimes.

**Reply by textbrowser** [reply]
Are you retired?

**Comment by textbrowser** [comment]
That's a lengthy article. It covers the etiquette and art of problem reporting.

**Reply by meteos77** [reply]
we can make a report but I am not competent enough to say that it is a problem.
the proof is that it was not a problem with the software
for me the blue hand is a user asking for help because he has a user problem

**Reply by textbrowser** [reply]
Yup.

**Comment by meteos77** [comment, marked as answer]
Summary of response

a utf8 file can encode accented characters in several ways.
- if they are encoded on 1 character everything is fine
- if they are encoded on several characters BiblioteQ displays the character correctly but the internal processing keeps the 2 characters.

So some SQL queries (for example SELECT SUBSTR) answer in unexpected ways.

solutions :
- when importing csv, check that the accented characters are on 1 character.
- modify the accented characters by the BiblioteQ interface

**Reply by textbrowser** [reply]
Good work. :)

**Reply by meteos77** [reply]
[a little bonus](https://www.win.tue.nl/~aeb/linux/uc/nfc_vs_nfd.html)

with this command my problem disappeared : convert from utf-8 NFD to utf-8 NFC

`uconv -f utf-8 -t utf-8 -x nfc source_file.csv -o destination_file.csv`


Thanks to BiblioteQ to discover other worlds :-)
